diff --git a/CMake/Utils/AndroidMacros.cmake b/CMake/Utils/AndroidMacros.cmake
index b3c9b8d34..78be6559f 100644
--- a/CMake/Utils/AndroidMacros.cmake
+++ b/CMake/Utils/AndroidMacros.cmake
@@ -85,21 +85,21 @@ macro(create_android_proj ANDROID_PROJECT_TARGET)
     configure_file("${OGRE_TEMPLATES_DIR}/AndroidManifest.xml.in" "${NDKOUT}/AndroidManifest.xml" @ONLY)
     file(WRITE "${NDKOUT}/default.properties" "target=${ANDROID_TARGET}")
     
-    if(EXISTS ${ANDROID_EXECUTABLE})
-        add_custom_command(
-            TARGET ${ANDROID_PROJECT_TARGET}
-            POST_BUILD
-            COMMAND ${ANDROID_EXECUTABLE} update project  --target 1 --path "${NDKOUT}"
-            WORKING_DIRECTORY ${NDKOUT}
-        )
+    # if(EXISTS ${ANDROID_EXECUTABLE})
+    #     add_custom_command(
+    #         TARGET ${ANDROID_PROJECT_TARGET}
+    #         POST_BUILD
+    #         COMMAND ${ANDROID_EXECUTABLE} update project  --target 1 --path "${NDKOUT}"
+    #         WORKING_DIRECTORY ${NDKOUT}
+    #     )
 
-        add_custom_command(
-            TARGET ${ANDROID_PROJECT_TARGET}
-            POST_BUILD
-            COMMAND ${ANT_EXECUTABLE} debug
-            WORKING_DIRECTORY ${NDKOUT}
-        )
-    else()
-        message(WARNING "Android executable not found. Not building ${ANDROID_PROJECT_TARGET} APK. Do you have the Android SDK installed?")
-    endif()
+    #     add_custom_command(
+    #         TARGET ${ANDROID_PROJECT_TARGET}
+    #         POST_BUILD
+    #         COMMAND ${ANT_EXECUTABLE} debug
+    #         WORKING_DIRECTORY ${NDKOUT}
+    #     )
+    # else()
+    #     message(WARNING "Android executable not found. Not building ${ANDROID_PROJECT_TARGET} APK. Do you have the Android SDK installed?")
+    # endif()
 endmacro(create_android_proj)
diff --git a/Samples/Browser/CMakeLists.txt b/Samples/Browser/CMakeLists.txt
index 5d68b6aac..75c43ecf4 100644
--- a/Samples/Browser/CMakeLists.txt
+++ b/Samples/Browser/CMakeLists.txt
@@ -157,7 +157,7 @@ target_link_libraries(SampleBrowser OgreMain ${OGRE_PLUGIN_LIBRARIES} ${SAMPLE_L
 
 if(ANDROID)
     target_link_libraries(SampleBrowser native_app_glue cpufeatures android c m dl z log EGL GLESv2 "-u ANativeActivity_onCreate")
-    set_target_properties(SampleBrowser PROPERTIES LINK_FLAGS -s) # strip symbols
+    # set_target_properties(SampleBrowser PROPERTIES LINK_FLAGS -s) # strip symbols
 endif()
 
 # Add samples as dependencies
